internal/controllers/preflightvalidation_reconciler.go

```diff
@@ -0,0 +1,249 @@
+
+	// we want the metric just to signal that this customer uses preflight
+	r.metricsAPI.SetKMMPreflightsNum(1)
+
+	modsWithMapping, modsWithoutMapping, err := r.helper.getModulesData(ctx, pv)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getModulesData failed: %v", err)
+	}
+
+	err = r.helper.updateStatus(ctx, modsWithMapping, modsWithoutMapping, pv)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("updateStatus failed: %v", err)
+	}
+
+	err = r.helper.runPreflightValidation(ctx, modsWithMapping, pv)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("runPreflightValidation failed: %v", err)
+	}
+
+	reconCompleted := r.preflightAPI.AllModulesVerified(pv)
+	if reconCompleted {
+		log.Info("PreflightValidation reconciliation success")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+//go:generate mockgen -source=preflightvalidation_reconciler.go -package=controllers -destination=mock_preflightvalidation_reconciler.go preflightReconcilerHelper
+type preflightReconcilerHelper interface {
+	updateStatus(ctx context.Context, modsWithMapping []*api.ModuleLoaderData, modsWithoutMapping []types.NamespacedName, pv *v1beta2.PreflightValidation) error
+	getModulesData(ctx context.Context, pv *v1beta2.PreflightValidation) ([]*api.ModuleLoaderData, []types.NamespacedName, error)
+	runPreflightValidation(ctx context.Context, modsWithMapping []*api.ModuleLoaderData, pv *v1beta2.PreflightValidation) error
+}
+
+type preflightReconcilerHelperImpl struct {
+	client       client.Client
+	micAPI       mic.MIC
+	metricsAPI   metrics.Metrics
+	kernelAPI    module.KernelMapper
+	preflightAPI preflight.PreflightAPI
+}
+
+func newPreflightReconcilerHelper(client client.Client,
+	micAPI mic.MIC,
+	metricsAPI metrics.Metrics,
+	kernelAPI module.KernelMapper,
+	preflightAPI preflight.PreflightAPI) preflightReconcilerHelper {
+
+	return &preflightReconcilerHelperImpl{
+		client:       client,
+		micAPI:       micAPI,
+		metricsAPI:   metricsAPI,
+		kernelAPI:    kernelAPI,
+		preflightAPI: preflightAPI,
+	}
+}
+
+func (p *preflightReconcilerHelperImpl) updateStatus(
+	ctx context.Context,
+	modsWithMapping []*api.ModuleLoaderData,
+	modsWithoutMapping []types.NamespacedName,
+	pv *v1beta2.PreflightValidation) error {
+
+	log := ctrl.LoggerFrom(ctx).WithValues("preflight name", pv.Name)
+	unmodifiedPV := pv.DeepCopy()
+
+	// setting the status for modules without mapping
+	for _, mod := range modsWithoutMapping {
+		p.preflightAPI.SetModuleStatus(pv, mod.Namespace, mod.Name, v1beta2.VerificationFailure, "mapping not found")
+	}
+
+	// setting status for modules with mapping
+	for _, mod := range modsWithMapping {
+		modStatus := v1beta2.VerificationInProgress
+		modReason := "verification is not finished yet"
+		log.Info("YEV - checking mic", "name", mod.Name+"-preflight", "namespace", mod.Namespace)
+		foundMIC, err := p.micAPI.Get(ctx, mod.Name+"-preflight", mod.Namespace)
+		if err == nil {
```

Comment:
How about we `continue` if we got an error instead of having this big `if` block?

Reply:
Sorry. I missed that.